Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -0,0 +1,780 @@
+			rotation = targetRotation;
+
+			return false;
+		}
+
+		bool ManipulatorSnapToSurface(Transform rayOrigin, ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, SnappingState state, Quaternion targetRotation, float breakDistance)
+		{
+			var bounds = state.identityBounds;
+			var boundsExtents = bounds.extents;
+			var projectedExtents = Vector3.Project(boundsExtents, Vector3.down);
+			var offset = projectedExtents - bounds.center;
+			var rotationOffset = Quaternion.AngleAxis(90, Vector3.right);
+			var startRotation = state.startRotation;
+			var upVector = startRotation * Vector3.back;
+			var maxRayLength = k_SurfaceSnappingMaxRayLength * this.GetViewerScale();
+
+			var pointerRay = new Ray(rayOrigin.position, rayOrigin.forward);
+			return SnapToSurface(pointerRay, ref position, ref rotation, state, offset, targetPosition, targetRotation , rotationOffset, upVector, m_IgnoreList, breakDistance, maxRayLength)
+				|| TryBreakSurfaceSnapping(ref position, ref rotation, targetPosition, startRotation, state, breakDistance);
+		}
+
+		bool DirectSnapToSurface(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, SnappingState state, Quaternion targetRotation, float breakDistance)
+		{
+			var bounds = state.identityBounds;
+			var boundsCenter = bounds.center;
+			for (int i = 0; i < k_Directions.Length; i++)
+			{
+				var direction = k_Directions[i];
+				var upVector = targetRotation * direction.upVector;
+				var directionVector = direction.direction;
+				var rotationOffset = direction.rotationOffset;
+				var boundsRay = new Ray(targetPosition + targetRotation * boundsCenter, targetRotation * directionVector);
+
+				var boundsExtents = bounds.extents;
+				var projectedExtents = Vector3.Project(boundsExtents, directionVector);
+				var raycastDistance = projectedExtents.magnitude * k_DirectSurfaceSearchScale;
+				var offset = -boundsCenter;
+				if (i > 2)
+					offset -= projectedExtents;
+				else
+					offset += projectedExtents;
+
+				if (SnapToSurface(boundsRay, ref position, ref rotation, state, offset, targetPosition, targetRotation, rotationOffset, upVector, m_IgnoreList, breakDistance, raycastDistance))
+					return true;
+			}
+
+			if (TryBreakSurfaceSnapping(ref position, ref rotation, targetPosition, targetRotation, state, breakDistance))
+				return true;
+
+			return false;
+		}
+
+		static bool TryBreakSurfaceSnapping(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, Quaternion targetRotation, SnappingState state, float breakDistance)
+		{
+			if (state.surfaceSnapping)
+			{
+				if (Vector3.Distance(position, targetPosition) > breakDistance)
+				{
+					position = targetPosition;
+					rotation = targetRotation;
+					state.surfaceSnapping = false;
+				}
+
+				return true;
+			}
+			return false;
+		}
+
+		static void SetupIgnoreList(List<GameObject> ignoreList, Renderer[] moduleIgnoreList, GameObject go)
+		{
+			ignoreList.Clear();
```

Comment:
Instead of copy/pasted code, can you keep a cache field for a local array of 1 element that you can set and use the other method below?